### DISCUSSION TITLE
FDC3 2.0 questions

### DISCUSSION THREAD

**Discussion by kriswest** [finos/FDC3#803] [General]
Please feel free to post FDC3 2.0 implementation, usage or capability questions in this discussion!

**Comment by novavi** [comment]
In the @finos/fdc3@2.0.0-beta.4 npm package, the DesktopAgent interface (src/api/DesktopAgent.ts) lists 23 methods in total, consisting of 17 current methods and 6 deprecated methods. As expected, those current and deprecated methods match the ones listed in the FDC3 2.0 docs under DesktopAgent (https://fdc3.finos.org/docs/api/ref/DesktopAgent).

Looking at the function signatures in src/api/Methods.ts, I notice they use union types in the args in order to specify 4 of the functions, in order to handle both current and deprecated signatures (i.e. open, raiseIntent, raiseIntentForContext, addContextListener). With this in mind, I was expecting to see 19 functions specified in src/api/Methods.ts i.e. the 23 total minus 4 which are represented only once (rather than twice) using union types in the args.

However, only 17 functions are actually specified in src/api/Methods.ts, rather than the 19 I was expecting. This is because createPrivateChannel and findInstances have been omitted.

@kriswest Was this intentional?

**Reply by kriswest** [reply]
@novavi no it wasn't. I opened [PR #876](https://github.com/finos/FDC3/pull/876) yesterday to add `createPrivateChannel`. I also had someone check for other missing functions, apparently, they missed the omission of `findInstances`. I'll go add that to the PR now, thanks for flagging.

FYI I believe the union types are required as the code is transpiled to javascript where there are no types on parameters. We can implement separate functions in TypeScript with its typing but I'm unsure what happens when its transpiled down... If anyone does know or is happy to explore I'd be happy to review and consider changes based on the result :-)

**Reply by kriswest** [reply]
P.S. the NPM module is still bearing a pre-release tag as it hasn't had a full review from the @finos/fdc3-maintainers yet and I'm not convinced it is yet free of errors - hence your input is much appreciated.

**Reply by kriswest** [reply]
PR updated: https://github.com/finos/FDC3/pull/876
needs maintainer review then will release

**Reply by kriswest** [reply]
The latest version with these fixes has been released @novavi  

https://www.npmjs.com/package/@finos/fdc3/v/2.0.0-beta.5

**Reply by novavi** [reply]
That's great, many thanks @kriswest 

As for the union types - yes I agree. The code is very clear, and I wasn't criticising the way the function signatures have been specified. I was only really making the observation that the manner in which TypeScript/JavaScript deals with overloading slightly clouded the issue at first glance when I eye-balled the number of methods in the interface in DesktopAgent.ts versus the number of function signatures in Methods.ts.

In terms of the approach taken in Methods.ts for overloading, this seems eminently sensible to me. As you alluded to, this is a fairly messy area in transpiled- (or indeed vanilla-) JavaScript because of the lack of support for types - and hence also overloading - at runtime. Vanilla JavaScript actually lets you create a class with two methods of the same name and with different signatures. But it's not a helpful feature - the final method simply overwrites the previous method implementation(s). This means you lose the overloads, and end up with just a single method. Worse, afaik it accepts this silently and doesn't even throw an error. This effectively makes it like monkey patching - but the twist being that it happens within the context of the single piece of code that defines the class, rather than by executing a subsequent piece of code to swap out the original class method. In practical terms, this is useless. I would hope not too many people are developing enterprise software in vanilla JavaScript nowadays anyway, so hopefully that behaviour is academic!

In terms of TypeScript support for overloading, as I understand it there's essentially two ways of doing it. The first way (along the lines of what is used in src/api/Methods.ts) is to have a single signature handle all the possible types and args by using a combination of union types (and, if need be, optional args). The second way (as used in the VS Code 'quick fix' to 'implement interface') involves defining multiple overload signatures without bodies, followed by a single implementation signature with a body (which is typically specified with 'unknown' arg types). The decision on this can be somewhat subjective, and arguably also influenced by the number of overloads required.

From an IntelliSense perspective though, in the case of a single signature with union types then I believe the developer will only be shown the single signature (with union types, etc.) instead of being shown multiple overload signatures from the interface that is being implemented. By contrast, when multiple overload signatures and a single implementation signature are used in a class then when the class is consumed, IntelliSense will show all the overload signatures.

**Reply by novavi** [reply]
Actually, what I said above was really more about how overloading works for class methods in TypeScript. The case of Methods.ts is subtly different, since it involves the export of function declarations rather than involving a class that is implementing an interface. I think I'm right in saying that TypeScript does not allow the export of multiple function implementations sharing the same name from a given source file, because they would be considered duplicates simply on the basis of the shared name alone. I suspect you would probably be able to export overloaded ambient function declarations from a type definition file but that's not so helpful in this case where you are dealing with function implementations in a code file.

I would need to check this in the latest version of TypeScript to be 100% sure, but based on what I believe is true about TypeScript from past experience then at least under the current structure of Methods.ts, having a single function signature handle all the possible types and args by using a combination of union types is probably the correct way to go.

Clearly, there are other possible ways of implementing this e.g. an exported class which has to be imported as a whole unit, rather than having exported functions available to the consumer for import on an individual basis (that would certainly be a more common pattern outside of the TypeScript/JavaScript world). However, I believe the way things are currently structured in Methods.ts is completely in line with current practice for TypeScript i.e. allowing the consumer to import only the functions they actually require, thus assisting the consumer's application build process and bundler to drop unused code from the final bundle based on an analysis of the dependency tree.

**Reply by novavi** [reply]
On reflection, I've had one other thought. I said above that when using classes in TypeScript, the choice of which mechanism to use to handle overloaded methods was somewhat subjective. And I suggested that since the export of function declarations in TypeScript _appears_ to push one down the path of having a single signature to handle all the possible types and args by using union types, then there is little downside to the current approach.

But there is another dimension to this. The second way I described in my comment yesterday for handling overloads in TypeScript classes (i.e. defining multiple overload signatures without bodies, followed by a single implementation signature with a body) does in fact have a clear benefit when it comes to the the presence of deprecated signatures. Having multiple overload signatures obviously allows the IDE to provide a visual clue with strikethrough when the developer attempts to use a deprecated signatures.

However, by using a single signature to handle all the possible types and args by using union types (as used in the current approach for the export of function declarations in Methods.ts) then unless I'm missing something it means that the developer will not see any strikethrough when attempting to use a deprecated signature. So perhaps this may actually be worth revisiting this at some stage after all?

**Reply by novavi** [reply]
Hi @kriswest my responses above were pretty theoretical, but I've chance now to investigate and try a couple of things - so I can now offer a better explanation and a proposed solution.

It looks like we can in fact have the best of both worlds here i.e. individually exported functions (which allow modern-style individual imports by the consuming application) and also benefit from the multiple signature support that we take for granted in TypeScript interfaces. I'd never tried this on functions before, but it turns out it works in the same manner as in TypeScript classes. Here's a working example for functions that I _believe_ could be used in `src/api/Methods.ts`:

```
export function addContextListener(contextType: string | null, handler: ContextHandler): Promise<Listener>;
/**
 * Adds a listener for incoming context broadcasts from the Desktop Agent.
 * @deprecated use `addContextListener(null, handler)` instead of `addContextListener(handler)`. Provided for backwards compatibility with versions FDC3 standard <2.0.
 */
export function addContextListener(handler: ContextHandler): Promise<Listener>;
export function addContextListener(
  contextTypeOrHandler: string | null | ContextHandler,
  handler?: ContextHandler
): Promise<Listener> {
  //Handle (deprecated) function signature that allowed contextType argument to be omitted
  if (typeof contextTypeOrHandler !== 'function') {
    return rejectIfNoGlobal(() => window.fdc3.addContextListener(contextTypeOrHandler, handler as ContextHandler));
  } else {
    return rejectIfNoGlobal(() => window.fdc3.addContextListener(null, contextTypeOrHandler as ContextHandler));
  }
}
```

From a IDE perspective, this means:

- When the consuming code is written, IntelliSense shows both the current and the deprecated signatures for the function, which makes it really clear which args you should pass in order to avoid using the deprecated version of the function.
- If the consuming code invokes the deprecated version of the function (i.e. by using the args that match the deprecated signature) then this is clearly indicated with a strikethrough on the function name.

Note however that if you mouseover your usage of the function, you will just see the single signature from the function implementation i.e. the one with the common argument list, which is the one which actually gets transpiled to JavaScript and executed at runtime.

Example consuming code for this would be:

```
// Using current version looks fine in the IDE
const listener = await addContextListener(ContextTypes.InstrumentList, (context: Context, metadata?: ContextMetadata): void => {
  console.log(JSON.stringify(context, null, 2));
});
    
// Using deprecated version clearly shows strikethrough in the IDE
const listener = await addContextListener((context: Context, metadata?: ContextMetadata): void => {
  console.log(JSON.stringify(context, null, 2));
});
```

In order update `src/api/Methods.ts` to provide a better developer experience, I _believe_ it should be as simple as copying the current and deprecated declarations from `src/api/DesktopAgent.ts` (including the `@deprecated` comment block as appropriate) and then adding these into `src/api/Methods.ts` immediately above the correspondly-named function implementation (and of course prefixing the declarations added with `export function`).

To be completely clear on what TypeScript supports in this area, these are things you cannot do:

- Export two same-named function implementations i.e ones that both have bodies.
- Mix ambient and non-ambient signatures for a given function name i.e. using both `export declare function` and `export function`.

And this is something you must do:
- When you declare the function declarations (i.e. those without bodies), they must be immediately followed by the same-named function implementation (i.e. that containing a body). You cannot have any other code between the two, and you cannot omit the function implementation - otherwise you will receive a "Function implementation is missing or not immediately following the declaration." error.

I trust the above will help in future updates / iterations of `@finos/fdc3`.

**Reply by kriswest** [reply]
Thanks for your write-up on this @novavi. I believe you are correct in that we *might* be able to improve the developer experience of using Methods.ts with the approach you outline (add the additional signatures without a body and eliminate the union types - particularly as FDC3 2.0 implementations are not *required* to support the deprecated signatures). 

Would you be up for submitting an issue (and/or PR) to have this resolved?

**Reply by novavi** [reply]
Hi @kriswest - no problem. Yes, I'm happy to submit an issue. And also happy to have a stab at a PR - though for the record I've spent my entire career in closed-source caves :) so I might need a little guidance around policies and procedures, if you're happy to provide.

**Reply by kriswest** [reply]
Hi @novavi contributions are much appreciated! You can find the contribution instructions at: https://github.com/finos/FDC3#contributing which links on to the contribution guide doc.

The main things to note are:
- raise an issue briefly describing the change needed and use-case
- raise a PR and link it to the issue (by saying "resolves #<issue num> in its description)
- The EasyCLA tool will spin up and check if it can find a contributor license agreement (CLA) for you - it probably won't the first time, but will give you a link to sort that out. If your firm has a Corporate CLA it may be able to validate you by email address, otherwise you're firm's CLA manager will have to add you to it. 
  - I believe @dominicgifford from your firm has submitted a PR before and has a CLA in place (as a proof-of-concept)

**Reply by kriswest** [reply]
@novavi just stumbled across this for the first time in a while - still think you can improve Methods.ts?

**Reply by novavi** [reply]
@kriswest yes, that's a good idea! This has been sitting here for ages now, as I shifted my attention to #896 after the discussion we had above and never returned to it.
